File: client/lib/shared/http/models/comment_model.dart
import 'package:client/shared/http/models/vote_model.dart';

import 'author_model.dart';

class CommentModel {
  final int id;
  final AuthorModel author;
  final String content;
  final List<VoteModel>? votes;
  final DateTime createdAt;
  final DateTime updatedAt;

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> json = <String, dynamic>{};
    json['id'] = id;
    json['author'] = author.toJson();
    json['content'] = content;
    json['createdAt'] = createdAt.toIso8601String();
    json['updatedAt'] = updatedAt.toIso8601String();
    json['votes'] = votes;

    return json;
  }

  CommentModel.fromJson(Map<String, dynamic> json)
      : id = json['id'],
        author = AuthorModel.fromJson(json['author']),
        content = json['content'],
        votes = json['votes']
            ?.map<VoteModel>((e) => VoteModel.fromJson(e))
            .toList(),
        createdAt = DateTime.parse(json['createdAt']),
        updatedAt = DateTime.parse(json['updatedAt']);
}

File: server/src/interfaces/CommentModel.ts
import { AuthorModel } from "./AuthorModel";
import { VoteModel, voteModelQuery } from "./VoteModel";

export const commentModelQuery = {
  id: true,
  author: {
    select: {
      id: true,
      name: true,
      picture: true,
    },
  },
  content: true,
  votes: {
    select: voteModelQuery,
  },
  createdAt: true,
  updatedAt: true,
};

export function _transformCommentModel(comment: any): CommentModel {
  return <CommentModel>{
    id: comment.id,
    author: comment.author,
    content: comment.content,
    votes: comment.votes,
    createdAt: comment.createdAt,
    updatedAt: comment.updatedAt,
  };
}

export interface CommentModel {
  id: number;
  author: AuthorModel;
  content: string;
  votes: VoteModel[];
  createdAt: Date;
  updatedAt: Date;
}
